vllm/core/scheduler.py

```diff
@@ -491,3 +495,18 @@ def _swap_out(
 
     def mark_blocks_as_computed(self, seq_group: SequenceGroup):
         self.block_manager.mark_blocks_as_computed(seq_group)
+
+    def _passed_delay(self, now: float) -> bool:
+        if self.prev_prompt:
+            self.last_prompt_latency = now - self.prev_time
+        self.prev_time, self.prev_prompt = now, False
+        # Delay scheduling prompts to let waiting queue fill up
+        if self.scheduler_config.use_delay and self.waiting:
+            earliest_arrival_time = min(
+                [e.metrics.arrival_time for e in self.waiting])
+            passed_delay = ((now - earliest_arrival_time) >
+                            (0.5 * self.last_prompt_latency)
+                            or not self.running)
+        else:
+            passed_delay = True
+        return passed_delay
```

Comment:
what happens if we have a long time without any new requests? isn't that going to mess up the calculation? ideally we'd get the time taken from the request output, but that will complicate the code and require communication from the engine...

Reply:
Interesting - presumably that could only happen if received a request with *only* a prompt step (e.g., single token generation), and then nothing else for a while, because as soon as we get a single decoding step, we will calculate the `last_prompt_latency`. Let me think about how to handle that.

I agree we could technically get the prompt latency from outside the scheduler, but as you say it will complicate the code. Ideally, it can be a behaviour self-contained within the scheduler.

Reply:
Thinking about it a little more, I think that case would be fine because if it happens then the `self.running` queue would also be empty which would mean that we don't enforce the delay.